core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -124,162 +136,170 @@ static Builder buildFrom(ViewMetadata base) {
+      this.properties = Maps.newHashMap();
+      this.changes = Lists.newArrayList();
+    }
 
     private Builder(ViewMetadata base) {
-      this.formatVersion = base.formatVersion();
-      this.location = base.location();
-      this.schemas = Lists.newArrayList(base.schemas());
-      this.currentVersionId = base.currentVersionId();
       this.versions = Lists.newArrayList(base.versions());
+      this.versionsById = Maps.newHashMap(base.versionsById());
+      this.schemas = Lists.newArrayList(base.schemas());
+      this.schemasById = Maps.newHashMap(base.schemasById());
       this.history = Lists.newArrayList(base.history());
       this.properties = Maps.newHashMap(base.properties());
-      this.changes = Lists.newArrayList(base.changes());
+      this.changes = Lists.newArrayList();
+      this.formatVersion = base.formatVersion();
+      this.currentVersionId = base.currentVersionId();
+      this.location = base.location();
     }
 
     public Builder upgradeFormatVersion(int newFormatVersion) {
       Preconditions.checkArgument(
-          newFormatVersion >= this.formatVersion,
+          newFormatVersion >= formatVersion,
           "Cannot downgrade v%s view to v%s",
           formatVersion,
           newFormatVersion);
 
-      if (this.formatVersion == newFormatVersion) {
+      if (formatVersion == newFormatVersion) {
         return this;
       }
 
       this.formatVersion = newFormatVersion;
-      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
       return this;
     }
 
     public Builder setLocation(String newLocation) {
       Preconditions.checkArgument(null != newLocation, "Invalid location: null");
-      if (null != this.location && this.location.equals(newLocation)) {
+      if (null != location && location.equals(newLocation)) {
         return this;
       }
 
       this.location = newLocation;
-      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      changes.add(new MetadataUpdate.SetLocation(newLocation));
       return this;
     }
 
     public Builder setCurrentVersionId(int newVersionId) {
-      if (newVersionId == -1) {
+      if (newVersionId == LAST_ADDED) {
         ValidationException.check(
             lastAddedVersionId != null,
             "Cannot set last version id: no current version id has been set");
         return setCurrentVersionId(lastAddedVersionId);
       }
 
-      if (this.currentVersionId == newVersionId) {
+      if (currentVersionId == newVersionId) {
         return this;
       }
 
-      checkCurrentVersionIdIsValid(newVersionId);
-
-      // by that time, schemas are defined and the highestFieldId can be determined and schema
-      // changes can be added
-      int highestFieldId = highestFieldId();
-      for (Schema schema : schemas) {
-        this.changes.add(new MetadataUpdate.AddSchema(schema, highestFieldId));
```

Comment:
The `AddSchema` change only needs the highest field ID with respect to the existing schemas and the new schema, so these changes can be added in `addSchema`.